File: Stacks/Stack.py
"""
A Stack is a LIFO (Last in first out) data structure.
Stack class using array

Time: O(1) for insertion and removal of items
Space: O(n)
"""


class Stack:
    """
    A simple implementation of a LIFO stack.
    """

    def __init__(self):
        """ 
        Initialize the stack.
        """
        self._items = []

    def __len__(self):
        """
        Return the number of items in the stack.
        """
        return len(self._items)

    def __iter__(self):
        """
        Create an iterator for the stack.
        """
        for item in self._items:
            yield item

    def __str__(self):
        """
        Return a string representation of the stack.
        """
        return str(self._items)

    def put(self, item):
        """
        Add item to the stack.
        """
        self._items.append(item)

    def pop(self):
        """
        Remove and return the most recently inserted item.
        """
        return self._items.pop()

    def clear(self):
        """
        Remove all items from the stack.
        """
        self._items = []


# Tests

stack = Stack()
print(len(stack))
stack.put(1)
stack.put(2)
stack.put(3)
stack.put(4)
print(len(stack))

# using generator function to loop through items in stack
for item in stack:
    print(item)

stack.pop()
print(len(stack))

for item in stack:
    print(item)

File: Stacks/stack.js
// A stack is a LIFO (Last in first out) data structure
// Useful for tracking a history of steps i.e. browsing history
// Stack implementation using a linked list

//Time: O(1) -> for push and pop operations
// Space: O(n)

class Node {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

class Stack {
    constructor() {
        this.top = null;
        this.size = 0;
    }

    push(val) {
        if (this.size === 0) {
            this.top = new Node(val);
        } else {
            const pushNode = new Node(val);
            pushNode.next = this.top;

            this.top = pushNode;
        }
        this.size++;
    }

    pop() {
        if (this.size === 0) return null;
        const poppedNode = this.top;
        this.top = this.top.next;
        this.size--;
        return poppedNode.val;
    }

    getTop() {
        return this.top.val;
    }

    to_string() {
        let str = "";
        let curr = this.top;
        while (curr !== null) {
            str += curr.val + ' -> ';
            curr = curr.next;
        }
        console.log(str);
    }
}


// Test
const myStack = new Stack();
myStack.push("1");
myStack.push("2");
myStack.push("3");
myStack.push("4");

console.log(myStack.to_string());
console.log(myStack.getTop());
console.log(myStack.size);
console.log(myStack.pop());
console.log(myStack.size);
console.log(myStack.getTop());
console.log(myStack.to_string());